### PR TITLE
Separate MurmurHash32 from hash_functions.cuh 

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Moves the `MurmurHash32` class definition from `hash_functions.cuh` to a new `murmur32.cuh` file.
Also moves the new file and the `hash_functions.cuh` from `cpp/include/cudf/detail/utilities/` to `cpp/include/cudf/hashing/detail`
The hash functions were redeclared from the `cudf::detail` namespace to the `cudf::hashing::detail` namespace.

The remaining changes are side-effects of making the above changes.
This PR is a follow on to PR #13626 
No new logic or functions have changed. Only internal detail headers have been refactored.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/hashing/detail/hash_functions.cuh
Comment (davidwendt) [inline cpp/include/cudf/hashing/detail/hash_functions.cuh:1]: This file was moved from `cpp/include/cudf/detail/utilities/hash_functions.cuh` and also split up
The `MurmurHash3_32` was moved to `murmurh_hash32.cuh`
Some utilities were only used by md5 and were moved into `md5_hash.cu`

File: cpp/include/cudf/hashing/detail/murmur_hash32.cuh
Comment (davidwendt) [inline cpp/include/cudf/hashing/detail/murmur_hash32.cuh:1]: This was moved from `cpp/include/cudf/detail/utilities/hash_functions.cuh`

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: cpp/src/hash/md5_hash.cu
Comment (davidwendt) [inline cpp/src/hash/md5_hash.cu:45]: All of this was moved from `cpp/include/cudf/detail/utilities/hash_functions.cuh`

File: cpp/include/cudf/hashing/detail/hash_functions.cuh
Comment (bdice) [inline cpp/include/cudf/hashing/detail/hash_functions.cuh:25]:
```suggestion
template <typename Key>
```

File: cpp/include/cudf/hashing/detail/murmur_hash32.cuh
Comment (bdice) [inline cpp/include/cudf/hashing/detail/murmur_hash32.cuh:1]: Let’s name this identically to the hash function, murmurhash3_32.cuh.

File: cpp/include/cudf/hashing/detail/murmur_hash32.cuh
Comment (bdice) [inline cpp/include/cudf/hashing/detail/murmur_hash32.cuh:36]: Should we note whether this is the x64 or x86 version? Should this be named MurmurHash3_x64_32 or similar? Clarity here is crucial since we are adding new variants in a separate PR.

File: cpp/include/cudf/join.hpp
Comment (bdice) [inline cpp/include/cudf/join.hpp:39]: Should internal uses of hash functions like this use the MurmurHash3 explicitly or should they use the “default” hash function?

File: cpp/src/hash/murmur_hash.cu
Comment (bdice) [inline cpp/src/hash/murmur_hash.cu:62]: As above, let’s pick a consistent name and choice of where underscores go. I understand we may have case differences between a class and function but the name and placement of underscores should be matched.

File: cpp/include/cudf/hashing/detail/hash_functions.cuh
Comment (bdice) [inline cpp/include/cudf/hashing/detail/hash_functions.cuh:29]: It’s awkward that this forward declaration forces users to include the MurmurHash3 header just to use the “default.” It injects a requirement of knowing the default so that the header can be included. I wonder if we can reorganize this so that there’s a utility header separate from the header defining the default hash functor. Then we can change the includes so that including the default hash header will include the MurmurHash3 header, which includes the utility header.

File: cpp/include/cudf/hashing/detail/murmur_hash32.cuh
Comment (davidwendt) [reply cpp/include/cudf/hashing/detail/murmur_hash32.cuh:36]: There is no x64 version if that helps.
https://github.com/aappleby/smhasher/blob/61a0530f28277f2e850bfc39600ce61d02b518de/src/MurmurHash3.cpp#L94

File: cpp/include/cudf/hashing/detail/default_hash.cuh
Comment (bdice) [inline cpp/include/cudf/hashing/detail/default_hash.cuh:24]: Is this forward declaration still meaningful now that the header defining `MurmurHash3_x86_32<Key>` is included?

File: cpp/include/cudf/hashing/detail/hashing.hpp
Comment (bdice) [inline cpp/include/cudf/hashing/detail/hashing.hpp:1]: I could see an argument to move the `hash_combine` utility functions like from this file into `hashing/detail/hash_functions.cuh`. What do you think?

File: cpp/src/hash/spark_murmurhash3_x86_32.cu
Comment (bdice) [inline cpp/src/hash/spark_murmurhash3_x86_32.cu:39]: Probably need an underscore after "Spark" to match the function name `spark_murmurhash3_x86_32`.
```suggestion
struct Spark_MurmurHash3_x86_32 {
```

File: cpp/src/partitioning/partitioning.cu
Comment (bdice) [inline cpp/src/partitioning/partitioning.cu:730]: This sentence is missing something to be grammatically correct. Maybe we can simplify it like:
```suggestion
 * @brief  This hash function simply returns the input value cast to the
 * result_type of the functor.
```

**Review by bdice — APPROVED** [review]
Approving with minor questions/suggestions.

File: cpp/include/cudf/hashing/detail/hashing.hpp
Comment (davidwendt) [reply cpp/include/cudf/hashing/detail/hashing.hpp:1]: Some of these are used by .cpp files so I did not want them to include `hash_functions.cuh` which has some CUDA intrinsics in it.

File: cpp/include/cudf/join.hpp
Comment (davidwendt) [reply cpp/include/cudf/join.hpp]: This turned out to be more involved so I will fix this in a follow-up PR.

File: cpp/src/hash/md5_hash.cu
Comment (karthikeyann) [reply cpp/src/hash/md5_hash.cu:45]: nit: consider renaming `md5_hash.cu` to `md5.cu` or `md5hash.cu` to be consistent with other files. ignore if present name itself is okay.

**Review by karthikeyann — APPROVED** [review]
Looks good to me.
Only couple of minor questions to be clarified.

File: cpp/src/partitioning/partitioning.cu
Comment (davidwendt) [reply cpp/src/partitioning/partitioning.cu:735]: Probably because of the `m_seed` but since that has an initial value it should be possible. 
Is there some benefit to declare the default constructor is `constexpr` ?

Comment (karthikeyann) [reply cpp/src/partitioning/partitioning.cu:735]: This question came up because other constructor and member functions are constexpr, and the default constructor is not.

Comment (karthikeyann) [reply cpp/src/partitioning/partitioning.cu:735]: `m_seed` may not be needed, since it's not used. can `m_seed` be removed?

File: cpp/include/cudf/hashing/detail/hashing.hpp
Comment (bdice) [reply cpp/include/cudf/hashing/detail/hashing.hpp:1]: FWIW Those CUDA intrinsics for funnel shifts don't actually compile to fewer instructions. Originally I thought it would, but I checked this a while back and didn't see any change in the PTX/SASS. I just hadn't gone back to replace those. If it helps, we can replace the intrinsics with the equivalent shifting code.

Comment (davidwendt) [reply cpp/include/cudf/hashing/detail/hashing.hpp:1]: We can address this in a follow-up PR.

**Comment by ttnghia** [comment]
Re-label as breaking as it causes breaking in spark-rapids-jni.

**Pull request merged**
